### DISCUSSION TITLE
How to upload files in FormData as Stream

### DISCUSSION THREAD

**Discussion by ugonight** [nodejs/undici#1638] [Q&A] [answered]
Hello undici team. 

I am planning to create a program to upload files on Node.js using undici.
What is the best way to send FormData data in stream?

The following code will send the file data as the text `[object Object]`.

```js
import { request, FormData, File } from 'undici'
import fs from 'fs'

let formData = new FormData()
// formData.append('file', new File([fs.readFileSync('test.mp4')], 'test.mp4')) // no stream
formData.append('file', fs.createReadStream('test.mp4'))

let res = await request('http://localhost:3000/', { method: 'POST', body: formData })
console.log(await res.body.text())
```

Thanks in advance.

**Comment by jimmywarting** [comment, marked as answer]
for this end we need to be able to get Blob/Files backed up by the filesystem.
there is a feature request for this here: https://github.com/nodejs/node/issues/37340 (give it a 👍)

anyhow... for now i think your best solution is to either 
1. read the hole file in to memory and construct a `new File` (like your uncommented line) (not recommended for large files)
2. use something like [fetch-blob](https://github.com/node-fetch/fetch-blob) to get Blob/Files that are actually backed up by the file system. (Undici support them as well if it acts like a duck and quack like a duck)

```js
import { fileFromSync } from 'fetch-blob/from.js'

// will only stat the file for size & last modified date
const file = fileFromSync('test.mp4', 'video/mp4')
// this will not actually read anything into memory (yet) only when it needs to upload/serialize the FormData
// will it actually start to read the file
formData.append('file', file)

**Reply by ugonight** [reply]
Thank you for the detailed explanation. Thanks to you, I now have a better understanding of the system.
I needed to do a large file upload, so I tried the fetch-blob method and it worked like a charm. Thank you very much.
